### PR TITLE
Update locidex merge version

### PR DESCRIPTION
**Pull request by sgsutcliffe**
First, we want update the locidex version used by gasnomenclature to [0.3.0](https://pypi.org/project/locidex/0.3.0/). The reason locidex was updated was to integrate the functionality of gasnomenclature module `input_assure` into locidex to address the user-story:

## STRY0016854: Optimize "input_assure" Code in Gasnomenclature Pipeline

As a Developer, I want to move the "input_assure" code from a Python script in the Gasnomenclature pipeline to become part of the "locidex merge" process, specifically the portion enabling overriding the sample names stored in the MLST JSON file, so that the pipeline becomes more efficient by reducing redundant I/O.

## Acceptance Criteria

- [x] The "input_assure" code should be successfully moved to the "locidex merge" process without causing any disruptions in the pipeline.
- [x] The "input_assure" code should maintain its functionality of enabling overriding the sample names stored in the MLST JSON file.
- [x] The Gasnomenclature pipeline should be updated to remove the "input_assure" step.

## Additional PR checklist:

- [x] Add tests
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated.

### CODE REVIEW

**Comment by sgsutcliffe** [comment]
In order for tests to pass, so that I wouldn't be bogged down in fixing the nf-tests I added a little conditional statement so that the  error report would only be made if it wasn't empty, but I will remove later:
```
        if ((\$(wc -l ${input_error} | cut -f1 -d" ") > 1)); then
            mv "${input_error}" ${combined_dir}/MLST_error_report_${combined_dir}.csv
        fi
```

**Comment by sgsutcliffe** [comment]
Removed the 'Check if error report is empty before saving' which broke a lot of nf-tests that compared the `iridanext.output.json`. This comparison broke, in a way that caused flakey tests, because the `global` key had values (paths to error reports) in different orders each time. I think some of the tests can be updated, as we should only need to check the error reports if for the tests that have an error report that is not empty. For now I added it for any test that had a `iridanext.output.json` compare.

**Comment by emarinier** [comment]
We're probably going to want to remove all linting warnings except for the `nf-validation`-related one.

File: conf/modules.config [hunk 1]
```diff
@@ -47,6 +46,25 @@ process {
         ]
     }
 
+    withName: LOCIDEX_CONCAT_REF {
+        publishDir = [
+            path: locidex_concat_ref_directory_name,
+            mode: params.publish_dir_mode,
+            pattern:  "*/*",
```
Comment (emarinier) [inline conf/modules.config:53]: What are the `*/*` patterns doing? I don't think I've ever used a line like this before.

File: modules/local/locidex/concat/main.nf [hunk 2]
```diff
@@ -8,27 +8,36 @@ process LOCIDEX_CONCAT {
         'biocontainers/csvtk:0.30.0--h9ee0642_0' }"
 
     input:
-    path(input_values, stageAs: "?/*") // [file(sample1), file(sample2), file(sample3), etc...]
-    val input_tag // makes output unique and denotes the item as the reference or query to preven name collision
+    path(input_profile, stageAs: "?/*") // [profiles1.tsv, profiles2.tsv, ..., profilesn.tsv]
+    path(input_error,   stageAs: "?/*") // [error_report1.csv, error_report2.csv, ... , error_reportn.csv]
```
Comment (emarinier) [inline modules/local/locidex/concat/main.nf:12]: I dont think I've ever seen these `?/` either. What do they do?

File: modules/local/locidex/concat/main.nf [hunk 3]
```diff
@@ -8,27 +8,36 @@ process LOCIDEX_CONCAT {
...
+    val input_tag                       // makes output unique and denotes the item as the reference or query to preven name collision
```
Comment (emarinier) [inline modules/local/locidex/concat/main.nf:13]: preven -> prevent

File: modules/local/locidex/merge/main.nf [hunk 4]
```diff
@@ -6,24 +6,24 @@ process LOCIDEX_MERGE {
...
-    task.ext.override_configured_container_registry != false ? 'docker.io/mwells14/locidex:0.2.3' :
-    'mwells14/locidex:0.2.3' }"
+        'https://depot.galaxyproject.org/singularity/locidex%3A0.3.0--pyhdfd78af_0' :
+        'biocontainers/locidex:0.3.0--pyhdfd78af_0' }"
+
 
     input:
     path input_values // [file(sample1), file(sample2), file(sample3), etc...]
-    val input_tag // makes output unique and denotes the item as the reference or query to preven name collision
+    val  input_tag    // makes output unique and denotes the item as the reference or query to preven name collision
```
Comment (emarinier) [inline modules/local/locidex/merge/main.nf:15]: preven -> prevent

File: workflows/gas_nomenclature.nf [hunk 5]
```diff
@@ -92,49 +94,77 @@ workflow GAS_NOMENCLATURE {
             }
             // Add the ID to the set of processed IDs
             processedIDs << meta.id
-
-            tuple(meta, mlst_file)}
-
+            //createOverrideFile.append("\n$meta.id,$mlst_file")
```
Comment (emarinier) [inline workflows/gas_nomenclature.nf:97]: Generally recommend removing commented out code.

File: workflows/gas_nomenclature.nf [hunk 6]
```diff
@@ -92,49 +94,77 @@ workflow GAS_NOMENCLATURE {
...
-    combined_references = LOCIDEX_CONCAT_REF(references.combined_profiles.collect(), ref_tag, references.combined_profiles.collect().flatten().count())
-    combined_queries = LOCIDEX_CONCAT_QUERY(queries.combined_profiles.collect(), query_tag, queries.combined_profiles.collect().flatten().count())
+    // LOCIDEX Step 2:
+    // Combine outputs
+
+    // LOCIDEX Concatenate References
+    combined_references = LOCIDEX_CONCAT_REF(references.combined_profiles.collect(),
+    references.combined_error_report.collect(),
+    ref_tag,
+    references.combined_profiles.collect().flatten().count())
```
Comment (emarinier) [inline workflows/gas_nomenclature.nf:161]: Might want to tab these in for visual clarity, if the linter allows it.

File: workflows/gas_nomenclature.nf [hunk 7]
```diff
@@ -92,49 +94,77 @@ workflow GAS_NOMENCLATURE {
...
+
+    // LOCIDEX Concatenate References
+    combined_queries = LOCIDEX_CONCAT_QUERY(queries.combined_profiles.collect(),
+    queries.combined_error_report.collect(),
+    query_tag,
+    queries.combined_profiles.collect().flatten().count())
```
Comment (emarinier) [inline workflows/gas_nomenclature.nf:167]: Tabbed in if linter allows.

**Review by emarinier — CHANGES_REQUESTED** [review]
Thanks Steven! A few comments and questions.

Also, could you add a bit more to `tests/modules/locidex/main.nf.test`:

- Checking the contents of outputs, rather than just `process.success`
- One or more empty files with other empty files or files containing data

**Comment by sgsutcliffe** [comment]
> Thanks Steven! A few comments and questions.
> 
> Also, could you add a bit more to `tests/modules/locidex/main.nf.test`:
> 
> * Checking the contents of outputs, rather than just `process.success`
> * One or more empty files with other empty files or files containing data

I agree @emarinier I'd erred on the side of not reworking tests (thinking it outside the scope of the story). I was mainly doing a patch job to make tests pass with new outputs from the `locidex merge` change. But I agree it would be a good time to rework some of them. I'll revisit them, and see what I can improve. 

Related, and something I would like a second opinion on, is that we currently have collection of `iridanext_output.json` files made for the different tests but we don't really utilize them fully in the test (prior it was just an assert of `expected == outcome` which broke with new `files.global.paths` which I scrapped). I do like them, they act as nice references to check things, perhaps if a test doesn't catch it or explain the test better. Do you think I should also rework to include the `metadata` section of the JSON file too so that I utilize them more? That way we don't need to hardcode the expected values in them. 

Sorry that was a longer follow-up then I expected.

**Comment by emarinier** [comment]
> I agree @emarinier I'd erred on the side of not reworking tests (thinking it outside the scope of the story). I was mainly doing a patch job to make tests pass with new outputs from the `locidex merge` change. But I agree it would be a good time to rework some of them. I'll revisit them, and see what I can improve.

I think if we're expecting certain behaviour from locidex, then it makes sense to test those outputs specifically, even more so if locidex or related processes have changed recently.
 
> Related, and something I would like a second opinion on, is that we currently have collection of `iridanext_output.json` files made for the different tests but we don't really utilize them fully in the test (prior it was just an assert of `expected == outcome` which broke with new `files.global.paths` which I scrapped). I do like them, they act as nice references to check things, perhaps if a test doesn't catch it or explain the test better. Do you think I should also rework to include the `metadata` section of the JSON file too so that I utilize them more? That way we don't need to hardcode the expected values in them.

Speaking personally, I prefer having the asserts be specific and in the test themselves, not comparing whole files exactly, because the information is in multiple places. I think it reads better to see how the test is set up, and then see the expectations right there. It's even nice if the inputs are all defined there too.

Also, I think having exact file comparison asserts increases the risk of shortcutting-related errors being introduced into the tests themselves too. That is, someone runs the pipeline with that data, copies and pastes the output into the expected output file for the test, and isn't thinking too much about what it's actually supposed to be. I guess shortcutting can cause problems anywhere, but those are just my thoughts.

**Comment by sgsutcliffe** [comment]
> We're probably going to want to remove all linting warnings except for the `nf-validation`-related one.

Some seem pretty entrenched and I couldn't figure out how to remove them. I removed some df8596d
` nextflow_config: Config manifest.version should end in dev: 0.3.2 ` can only be fixed by running nf-core lint --release or adding `dev` to the end of the version which I don't think we want to do either. I couldn't find anything for `schema_lint` so I will have to leave that one in too.

File: conf/modules.config (hunk 1 above)
Comment (sgsutcliffe) [reply conf/modules.config]: 8d26092

File: modules/local/locidex/concat/main.nf (hunk 2 above)
Comment (sgsutcliffe) [reply modules/local/locidex/concat/main.nf:12]: `?` is a special wildcard that gives a number to each input (see [here](https://www.nextflow.io/docs/latest/process.html#input-type-path)). I probably didn't need to make it a directory for each I probably could have done like `profile_?.tsv` I just used the example [here](https://github.com/nf-core/modules/blob/3a1919641c75a83d8b12a43d907b318f694f04b0/modules/nf-core/biobambam/bamsormadup/main.nf#L9). It is useful for avoiding collisions.

File: modules/local/locidex/merge/main.nf (hunk 4 above)
Comment (sgsutcliffe) [reply modules/local/locidex/merge/main.nf]: 670c1ca

File: workflows/gas_nomenclature.nf (hunk 5 above)
Comment (sgsutcliffe) [reply workflows/gas_nomenclature.nf]: Missed that from testing! e99e98d

File: workflows/gas_nomenclature.nf (hunk 6 above)
Comment (sgsutcliffe) [reply workflows/gas_nomenclature.nf]: 49d787c

File: workflows/gas_nomenclature.nf (hunk 7 above)
Comment (sgsutcliffe) [reply workflows/gas_nomenclature.nf]: 49d787c

**Comment by sgsutcliffe** [comment]
Decided to only update the test for the `LOCIDEX_CONCAT` process nf-test 61c6673

File: modules/local/locidex/concat/main.nf (hunk 2 above)
Comment (emarinier) [reply modules/local/locidex/concat/main.nf:12]: Oh, interesting!

File: CHANGELOG.md [hunk 8]
```diff
@@ -3,6 +3,20 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.5.0] - 2025/03/dd
```
Comment (apetkau) [inline CHANGELOG.md:6]: Can you add the link the the anticipated release like the other versions at the bottom here?

File: tests/modules/locidex/main.nf.test [hunk 9]
```diff
@@ -23,6 +24,16 @@ nextflow_process {
 
         then {
             assert process.success
+
+            // Check the merged profiles files
+            def actual_profile = path("$baseDir/tests/data/profiles/expected-profile1.tsv")
```
Comment (apetkau) [inline tests/modules/locidex/main.nf.test:29]: I think you may have these variables mixed up. The `actual_profile` would be the one generated by the pipeline (in the `$launchDir` and the `expected_profile` would be the file we store in our repository test data. Could you switch the variable names?

File: workflows/gas_nomenclature.nf [hunk 10]
```diff
@@ -22,17 +23,17 @@ include { paramsSummaryLog; paramsSummaryMap; fromSamplesheet  } from 'plugin/nf
 // SUBWORKFLOW: Consisting of a mix of local and nf-core/modules
 //
 
-include { INPUT_ASSURE                          } from "../modules/local/input_assure/main"
-include { LOCIDEX_MERGE as LOCIDEX_MERGE_REF    } from "../modules/local/locidex/merge/main"
-include { LOCIDEX_MERGE as LOCIDEX_MERGE_QUERY  } from "../modules/local/locidex/merge/main"
+include { WRITE_METADATA                         } from '../modules/local/write/main'
```
Comment (apetkau) [inline workflows/gas_nomenclature.nf:26]: A small comment that isn't really necessary to change, but all other includes you use double quotes `"`. Could you change to double quotes here too?

File: nextflow.config [hunk 11]
```diff
@@ -181,7 +181,7 @@ process.ext.override_configured_container_registry = true
 // Nextflow plugins
 plugins {
     id 'nf-validation@1.1.3' // Validation of pipeline parameters and creation of an input channel from a sample sheet
-    id 'nf-iridanext@0.2.0' // Generation of JSON output for IRIDA Next
+    id 'nf-iridanext@0.3.0' // Generation of JSON output for IRIDA Next
```
Comment (apetkau) [inline nextflow.config:184]: Could you update the version number of the pipeline in this file as well.

**Review by apetkau — CHANGES_REQUESTED** [review]
This looks amazing @sgsutcliffe. Thanks so much for your great work on this 😄. I have a few comments.

File: CHANGELOG.md (hunk 8 above)
Comment (sgsutcliffe) [reply CHANGELOG.md:6]: a17ea4d

File: tests/modules/locidex/main.nf.test (hunk 9 above)
Comment (sgsutcliffe) [reply tests/modules/locidex/main.nf.test]: c705858

File: workflows/gas_nomenclature.nf (hunk 10 above)
Comment (sgsutcliffe) [reply workflows/gas_nomenclature.nf]: 312bb0b

File: nextflow.config (hunk 11 above)
Comment (sgsutcliffe) [reply nextflow.config:184]: 6a9ba14
